gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcOperationApi.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import com.google.api.core.ApiFunction;
+import com.google.api.gax.rpc.OperationApi;
+import com.google.api.gax.rpc.OperationSnapshot;
+import com.google.api.gax.rpc.TranslatingUnaryCallable;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.longrunning.CancelOperationRequest;
+import com.google.longrunning.DeleteOperationRequest;
+import com.google.longrunning.GetOperationRequest;
+import com.google.longrunning.Operation;
+import com.google.longrunning.stub.OperationsStub;
+import com.google.protobuf.Empty;
+
+public class GrpcOperationApi implements OperationApi {
+  private final OperationsStub operationsStub;
+
+  public GrpcOperationApi(OperationsStub operationsStub) {
+    this.operationsStub = operationsStub;
+  }
+
+  @Override
+  public UnaryCallable<String, OperationSnapshot> getOperationCallable() {
+    return TranslatingUnaryCallable.create(
+        operationsStub.getOperationCallable(),
+        new ApiFunction<String, GetOperationRequest>() {
+          @Override
+          public GetOperationRequest apply(String request) {
+            return GetOperationRequest.newBuilder().setName(request).build();
+          }
+        },
+        new ApiFunction<Operation, OperationSnapshot>() {
+          @Override
+          public OperationSnapshot apply(Operation operation) {
+            return GrpcOperationSnapshot.create(operation);
```

Comment:
So this allows to wrap Operation in OperationSnapshot, but I still don't understand what exactly problem is solved by OperationSnapshot? Allows to remvoe 4th parametrized parameter?

I still don't have good understanding on what is happening here (so sorry if the following does not make sense). It seems like `OperationApi + OperationSnapshot` are trying to build a manual-on-gapic client on `OperationsClient`, whithout actually doing one. Maybe it is better to wrap all that into a manual on top of gapic's `OperationsClient` (i.e. create some wraper around, which will return OperationSnapshot instead Operation). 

Also, TranslatingUnaryCallable comes with cost of ApiFutures.transform() on operation result. Those transforms are not as smooth as we would like them to be (they imply bidirectional (from wrapper to wrapped obj, and from wrapped obj to wrapper)) completion/failure/cancelation propagation. This is a very hard goal to achieve and it doesn't work under all circumstances (when there are unexpected errors in the middle, for example in transformers themselves). I would really like to avoid transforms if not absolutely necessary.

Reply:
The creation of `OperationSnapshot` is a consequence of creating `OperationApi` in core - you need some way to return the `Operation` data back. If you use a parameterized type, then you can't perform any logic on it, e.g. the polling logic that I'm moving up to core. 

Yes, this is kind of a manual-on-gapic client, but it also has to be possible for the interface to be backed by a non-gapic data source - for http1 APIs, this may look very different. 

I don't fully understand your concerns with the transformers. They are really just type transformers on the input/output types, and I didn't think they would affect completion/failure/cancelation propagation.